kernel/pci/src/lib.rs

```diff
@@ -217,33 +236,74 @@ pub fn pci_device_iter() -> Result<impl Iterator<Item = &'static PciDevice>, &'s
 }
 
 static INTX_DEVICES: Mutex<Vec<&'static PciDevice>> = Mutex::new(Vec::new());
-
-// Architecture-independent PCI interrupt handler
-// Currently aarch64-only, because legacy interrupts aren't supported on x86 yet.
-#[cfg(target_arch = "aarch64")]
-interrupt_handler!(pci_int_handler, None, _stack_frame, {
-    let devices = INTX_DEVICES.lock();
-
-    for device in &*devices {
-        if device.pci_get_interrupt_status(true) {
-            device.pci_enable_interrupts(false);
-            log::info!("Device {} triggered an interrupt", device.location);
-
-            let reader = device.interrupt_waker.lock();
-            match &*reader {
-                Some(waker) => waker.wake_by_ref(),
-                None => log::error!("Device doesn't have an interrupt waker!"),
+static INTX_NUMBERS: Mutex<[Option<InterruptNumber>; 4]> = Mutex::new([None; 4]);
+
+// Architecture-independent INTx handlers
+macro_rules! intx_handler {
+    ($name:ident, $num:literal) => {
+        interrupt_handler!($name, {
+            let intx_numbers = INTX_NUMBERS.lock();
+            intx_numbers[$num].expect("uninitialized x86 PCI INTx handler")
```

Comment:
This is a bit funky and kinda hard to read; I think this is a good place to just include this logic in the actual body of the interrupt handler and issue an EOI directly in the body instead of returning `HandlerDidNotSendEoi`. 

Another thing we could do is to include the interrupt number alongside the `waker` instance or alongside the `INTX_DEVICES` static, in order to "ensure" those are always available.